contracts/fast/FastTokenFacet.sol

```diff
@@ -249,92 +247,112 @@ contract FastTokenFacet is AFastFacet, IERC20, IERC1404 {
-    LibFast.Data storage d = LibFast.data();
-    ITokenHoldings(d.marketplace).holdingUpdated(p.from, address(this));
-    ITokenHoldings(d.marketplace).holdingUpdated(p.to, address(this));
+      // Keep track of who has what FAST.
+      LibFast.Data storage d = LibFast.data();
+      ITokenHoldings(d.marketplace).holdingUpdated(p.from, address(this));
+      ITokenHoldings(d.marketplace).holdingUpdated(p.to, address(this));
 
-    // Keep track of who holds this token.
-    holdingUpdated(p.from);
-    holdingUpdated(p.to);
+      // Keep track of who holds this token.
+      holdingUpdated(p.from);
+      holdingUpdated(p.to);
+    }
 
     // If the funds are not moving from the zero address, decrease transfer credits.
     if (p.from != address(0)) {
       // Make sure enough credits exist.
-      require(
-        s.transferCredits >= p.amount,
-        LibConstants.INSUFFICIENT_TRANSFER_CREDITS
-      );
-      s.transferCredits -= p.amount;
+      if (s.transferCredits < p.amount) {
+        return INSUFFICIENT_TRANSFER_CREDITS_CODE;
+      }
+      // Unless we're in dry-run mode, update state.
+      if (!dryRun) {
+        s.transferCredits -= p.amount;
+      }
     }
 
-    // If the funds are going to the ZERO address, decrease total supply.
-    if (p.to == address(0)) {
-      s.totalSupply -= p.amount;
-      // If funds at address zero changed, we can emit a top-level details change event.
-      FastFrontendFacet(address(this)).emitDetailsChanged();
-    }
-    // If the funds are moving from the zero address, increase total supply.
-    else if (p.from == address(0)) {
-      s.totalSupply += p.amount;
-      // If funds at address zero changed, we can emit a top-level details change event.
-      FastFrontendFacet(address(this)).emitDetailsChanged();
-    }
+    // Unless we're in dry-run mode, update state.
+    if (!dryRun) {
+      // If the funds are going to the ZERO address, decrease total supply.
+      if (p.to == address(0)) {
+        s.totalSupply -= p.amount;
+        // If funds at address zero changed, we can emit a top-level details change event.
+        FastFrontendFacet(address(this)).emitDetailsChanged();
+      }
+      // If the funds are moving from the zero address, increase total supply.
+      else if (p.from == address(0)) {
+        s.totalSupply += p.amount;
+        // If funds at address zero changed, we can emit a top-level details change event.
+        FastFrontendFacet(address(this)).emitDetailsChanged();
+      }
 
-    // Keep track of the transfer in the history facet.
-    FastHistoryFacet(address(this)).transfered(p.spender, p.from, p.to, p.amount, p.ref);
+      // Keep track of the transfer in the history facet.
+      FastHistoryFacet(address(this)).transfered(p.spender, p.from, p.to, p.amount, p.ref);
 
-    // Emit!
-    emit Transfer(p.from, p.to, p.amount);
+      // Emit!
+      emit Transfer(p.from, p.to, p.amount);
+    }
+
+    return 0;
```

Comment:
Clever... using `0` to signify it's ok (aka Unix land...) then return a specific uint8 for the code if the conditions don't pass. Using dryrun etc to ring fence state changing logic. Nice man 👍

Reply:
You're correct - my objective was to allow for a single source of algorythmic truth shared between `transfer` functions and `detectTransferRestriction` functions...